Added individual processor message telemetry events

akoutmos opened the pull request:
Creating a new PR to implement individual processor events per message (as was discussed in #152).

josevalim commented on lib/broadway/processor.ex:
```diff
@@ -117,16 +117,34 @@ defmodule Broadway.Processor do
       kind, reason ->
         reason = Exception.normalize(kind, reason, __STACKTRACE__)
 
+        emit_message_error_event(start_time, telemetry_metadata, reason)
```
The metadata for the error event should be: `kind: kind, reason: reason, stacktrace: __STACKTRACE__`. :)

akoutmos commented:
I think my most recent commit addresses the comments that you had. If there is anything else you would like me to tweak feel free to let me know. Thanks!

josevalim commented on lib/broadway/processor.ex:
```diff
@@ -168,17 +168,23 @@ defmodule Broadway.Processor do
+  defp emit_message_stop_event(start_time, metadata, handle_message_result) do
     stop_time = System.monotonic_time()
     measurements = %{time: stop_time, duration: stop_time - start_time}
+    updated_metadata = Map.put(metadata, :updated_message, handle_message_result)
```
Sorry for the nitpick but instead of doing `Map.put`, let's build the whole map, it will be easier to see what is the actual metadata. We can build the map either outside or pass all parameters to build the map inside.

akoutmos replied:
No reason to apologize :). I genuinely appreciate the feedback and think that your suggestion makes it more explicit as to what you can expect in the metadata map. Fixed in latest commit!

josevalim commented on lib/broadway/processor.ex:
```diff
@@ -168,17 +168,23 @@ defmodule Broadway.Processor do
+      metadata
+      |> Map.put(:error_kind, kind)
+      |> Map.put(:error_reason, reason)
+      |> Map.put(:error_stacktrace, stacktrace)
```
Same as above in regards to `Map.put`. Also, let's name the keys `kind`, `reason`, and `stacktrace`, which is the convention proposed by `telemetry`.

akoutmos replied:
Fixed in latest commit

josevalim commented:
Thanks @akoutmos! I have added two last comments. Also, please don't forget to update the docs with the new event. :)

akoutmos commented:
I went ahead and refactored the metadata maps as you suggested and also updated the docs to include the new events. I also noticed that in 081c1bb `batch_info` was added to the consumer start/stop events. I went ahead and added those entries to the consumer docs and also refactored the existing processor docs to make it clear how these per-message events differ.

Let me know if you want me to tweak/fix anything. Thanks!

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

akoutmos commented:
Thanks a bunch @josevalim !!
